### PR TITLE
8240640: [macos] Wrong focus behaviour with multiple Alerts

### PR DESCRIPTION
**Pull request by pankaj-bansal**
The bug is a regression as a result of fix done for JDK-8227366 and is reproducible on Linux and Mac. This fix is being reverted in this change and a new bug (JDK-8271054) has been created to redo the JDK-8227366

An automated testcase is being added to make sure similar regression is not introduced when working on the redo bug. The automated testcase fails without the current change and passes after the change.

The testcase works fine on Mac and Linux, but I see some issues in windows. The stage is minimised after calling stage.show without calling stage.setAlwaysOnTop(true). I see that there are other tests like DualWindowsTest.java which are failing for same reason. Also, if I remove the stage.setAlwaysOnTop(true) from tests like FocusParentWindowOnChildCloseTest.java,  the stage remains minimised there as well. This seems wrong as calling stage.setAlwaysOnTop(true) should not be required to bring stage window in focus on running the test. This may be an issue on my Window machine or something similar. I hope reviewers can shed some light on this.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8240640](https://bugs.openjdk.java.net/browse/JDK-8240640): [macos] Wrong focus behaviour with multiple Alerts

### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/581/head:pull/581` \
`$ git checkout pull/581`

Update a local copy of the PR: \
`$ git checkout pull/581` \
`$ git pull https://git.openjdk.java.net/jfx pull/581/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 581`

View PR using the GUI difftool: \
`$ git pr show -t 581`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/581.diff">https://git.openjdk.java.net/jfx/pull/581.diff</a>

</details>

### CODE REVIEW

File: tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            stage.setScene(new Scene(root, 500, 500));
+            stage.setOnShown(event -> Platform.runLater(startupLatch::countDown));
+            stage.show();
+
+            showAlert("Alert 1");
+            showAlert("Alert 2");
+            showAlert("Alert 3");
+        }
+
+        private void showAlert(String title)
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:104]: Minor: curly brace on same line as method.

File: tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            Alert alert = new Alert(Alert.AlertType.INFORMATION);
+            alert.initOwner(stage);
+            alert.setTitle(title);
+            alert.setOnShown(event -> Platform.runLater(startupLatch::countDown));
+            alert.setOnHidden(event -> Platform.runLater(alertCloseLatch::countDown));
+            alert.show();
+        }
+    }
+}
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:114]: Minor: add missing newline.

**Review by kevinrushforth — COMMENTED** [review]
The fix (reverting the earlier fix for JDK-8227366) looks fine to me. The test runs fine on Mac (and mostly fine on Linux, but I need to test a couple more things).

I also see the failure on Windows frequently depending on whether the Window initially gets focus. One thing you might try is to call `stage.toFront()` and `stage.requestFocus()` before showing the Alerts, although I'm not sure that will help. For other tests like this we do set the primary stage to be always on top, so that's probably the easiest. Failing that you could mark the test as unstable on Windows.

Btw, on my slow Linux VM, it occasionally fails because the window takes a bit too long to come up. I'll do some more testing once you fix the Windows issue.

I left a couple formatting comments on the test inline.

File: tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    @BeforeClass
+    public static void initFX() throws Exception {
+        new Thread(() -> Application.launch(TestApp.class, (String[]) null)).start();
+        waitForLatch(startupLatch, 15, "FX runtime failed to start.");
+    }
+
+    @Test(timeout = 25000)
+    public void testWindowFocusByClosingAlerts() throws Exception {
+        Thread.sleep(1000);
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:59]: On my slower Linux machine if I change this to `sleep(3000)` is passes reliably. Given that you are opening up multiple dialogs, this seems like a reasonable change.

File: tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java [hunk 4]
```diff
@@ -85,6 +89,16 @@ private static void keyPress(KeyCode code) throws Exception {
         });
     }
 
+    private static void mouseClick() {
+        Util.runAndWait(() -> {
+            robot.mouseMove((int) (alert.getX()+alert.getWidth()/2),
+                    (int) (alert.getY()+alert.getHeight()/2));
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:95]: Minor: we usually put spaces around binary operators.

**Review by kevinrushforth — APPROVED** [review]
The test changes look good (one minor formatting comment), and it now runs reliably for me on all three platforms.

**Comment by pankaj-bansal** [comment]
> The fix (reverting the earlier fix for JDK-8227366) looks fine to me. The test runs fine on Mac (and mostly fine on Linux, but I need to test a couple more things).
> 
> I also see the failure on Windows frequently depending on whether the Window initially gets focus. One thing you might try is to call `stage.toFront()` and `stage.requestFocus()` before showing the Alerts, although I'm not sure that will help. For other tests like this we do set the primary stage to be always on top, so that's probably the easiest. Failing that you could mark the test as unstable on Windows.
> 
> Btw, on my slow Linux VM, it occasionally fails because the window takes a bit too long to come up. I'll do some more testing once you fix the Windows issue.
> 
> I left a couple formatting comments on the test inline.

I tried the stage.toFront or stage.requestFocus, they did not work. So I added the stage.setAlwaysOnTop(true). Along with that, I also added functionality to do mouse click at centre of alert to make sure keyboard focus is present on the alert. I have verified that the updated test fails before the fix and passes with the fix. I have tried multiple iterations of the test on all platforms (Mac 10.15, Ubuntu 20.04 and Windows 10) and test looks stable.

File: tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java (hunk 4 above)
Comment (pankaj-bansal) [reply tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java]: Fixed

**Review by kevinrushforth — APPROVED** [review]

**Review by arapte — COMMENTED** [review]
looks good to me.

**Review by arapte — APPROVED** [review]

**Pull request closed without merging**